CSC8503CoreClasses/GuardObject.cpp

```diff
@@ -2,42 +2,175 @@
-	Ray r = Ray(this->GetTransform().GetPosition(), dir);
 	float ang = Vector3::Dot(dir, AngleOfSight());
 	if (ang > 2) {
+		RayCollision closestCollision;
+		Ray r = Ray(this->GetTransform().GetPosition(), dir);
 		if (mWorld->Raycast(r, closestCollision, true, this)) {
 			mSightedObject = (GameObject*)closestCollision.node;
 			Debug::DrawLine(this->GetTransform().GetPosition(), closestCollision.collidedAt);
 			if (mSightedObject == mPlayer) {
-				std::cout << "Gotem";
+				CanSeePlayer = true;
 			}
+			else {
+				CanSeePlayer = false;
+			}
+		}
+		else {
+			CanSeePlayer = false;
 		}
 	}
+	else {
+		CanSeePlayer = false;
+		mSightedObject = nullptr;
+	}
 }
 
 Vector3 GuardObject::AngleOfSight() {
-	Vector3 rightAxis = this->transform.GetMatrix().GetColumn(0);
+	Vector3 rightAxis = this->mTransform.GetMatrix().GetColumn(0);
 	Vector3 fwdAxis = Vector3::Cross(Vector3(0, 1, 0), rightAxis);
 	return fwdAxis;
+}
+
+void GuardObject::BehaviourTree() {
+
+	BehaviourSelector* FirstSelect = new BehaviourSelector("First Selector");
+	BehaviourSequence* CaughtPlayerSequence = new BehaviourSequence("Caught Player Sequence");
+	mRootSequence->AddChild(FirstSelect);
+	FirstSelect->AddChild(Patrol());
+	FirstSelect->AddChild(ChasePlayerSetup());
+	//FirstSelect->AddChild(CaughtPlayerSequence);
+	//CaughtPlayerSequence->AddChild(ConfiscateItems());
+	//CaughtPlayerSequence->AddChild(SendToPrison());
+
+}
+
+void GuardObject::ExecuteBT() {
+	BehaviourState state = mRootSequence->Execute(1.0f);
+	if (state == Success || state == Failure) {
+		mRootSequence->Reset();
+	}
+}
+
+BehaviourAction* GuardObject::Patrol() {
+	BehaviourAction* Patrol = new BehaviourAction("Patrol", [&](float dt, BehaviourState state)->BehaviourState {
+		if (state == Initialise) {
+			state = Ongoing;
+		}
+		else if (state == Ongoing) {
+			if (CanSeePlayer == false) {
+				std::cout << "Lost em";
+				return Success;
+			}
+			else if (CanSeePlayer == true) {
+				return Failure;
+			}
+		}
+		return state;
+	}
+	);
+	return Patrol;
+}
+
+BehaviourAction* GuardObject::ChasePlayerSetup() {
+	BehaviourAction* ChasePlayer = new BehaviourAction("Chase Player", [&](float dt, BehaviourState state)->BehaviourState {
+		if (state == Initialise) {
+			state = Ongoing;
+		}
+		else if (state == Ongoing){
+			if (CanSeePlayer == true && HasCaughtPlayer == false) {
+				Vector3 direction = mPlayer->GetTransform().GetPosition() - this->GetTransform().GetPosition();
+				this->GetPhysicsObject()->AddForce(Vector3(direction.x, 0, direction.z));
+				mPlayer = dynamic_cast<PlayerObject*>(mPlayer);
+				float dist = (direction.x * direction.x) + (direction.y * direction.y) + (direction.z * direction.z);
+				if (dist < 5) {
+					HasCaughtPlayer == true;
+					return Failure;
+				}
+			}
+			else if (CanSeePlayer == false && HasCaughtPlayer == false) {
+				return Success;
+			}
+			else if (CanSeePlayer == true && HasCaughtPlayer == true) {
+				return Failure;
```

Comment:
if this returns Failure anyway is there any point in this check? why not just let it fall through to else?